go: support "all:" prefix for "go:embed" directives

malt3 opened the pull request:
<!-- Thanks for sending a PR! Before submitting:

1. If this is your first PR, please read CONTRIBUTING.md and sign the CLA
   first. We cannot review code without a signed CLA.
2. Please file an issue *first*. All features and most bug fixes should have
   an associated issue with a design discussed and decided upon. Small bug
   fixes and documentation improvements don't need issues.
3. New features and bug fixes must have tests. Documentation may need to
   be updated. If you're unsure what to update, send the PR, and we'll discuss
   in review.
4. Note that PRs updating dependencies and new Go versions are not accepted.
   Please file an issue instead.
-->

**What type of PR is this?**

Feature

**What does this PR do? Why is it needed?**

Implements the "all:" prefix for `//go:embed` directives. From [Go docs](https://pkg.go.dev/embed#hdr-Directives):

> If a pattern begins with the prefix ‘all:’, then the rule for walking directories is changed to include those files beginning with ‘.’ or ‘_’. For example, ‘all:image’ embeds both ‘image/.tempfile’ and ‘image/dir/.tempfile’.

**Which issues(s) does this PR fix?**

Fixes #3441

**Other notes for review**

malt3 commented on tests/core/go_library/embedsrcs_test.go:
```diff
@@ -119,6 +122,22 @@ func TestFiles(t *testing.T) {
+			want: []string{
+				".",
+				"embedsrcs_static",
+				"embedsrcs_static/dir",
```
Had some trouble using `.git` as a folder name but I added some alternative examples

fmeum reviewed (approved):
Thanks, looks good!

On last thing: I just noticed that the Go docs only ever mention *files* starting with `.` and `_` - do you know whether `some/_hidden/file` would be included also without `all:...` with the standard Go toolchain?

malt3 commented:
> On last thing: I just noticed that the Go docs only ever mention files starting with . and _ - do you know whether some/_hidden/file would be included also without all:... with the standard Go toolchain?

The Go docs seem to be correct! Any directories are always included (including those starting with `.` or `_`). Should I fix the behavior of `rules_go` to follow this more closely?

fmeum commented:
That would be great, yes. Could you keep that change in a separate commit?

@linzhp @achew22 Do you have any backwards compatibility concerns?

malt3 commented:
A correct implementation should closely follow https://cs.opensource.google/go/go/+/master:src/cmd/go/internal/load/pkg.go;l=2045;drc=2f2c5e41e7728cae99abd14e8bd5ffcfebf9df8e

I do want to get there but I feel like we would need change the implementation more drastically to get it to be correct.

Additional edge cases I found:

- Hidden files can be conditionally in-/excluded based on if the folder name is given with a glob or not
  given the folder `a` and the file `a/.hidden`, the directive `//go:embed a` would exclude the file, while `//go:embed a/*` would include the file.
- There  are some exclude patterns for VCS: https://cs.opensource.google/go/go/+/master:src/cmd/go/internal/load/pkg.go;drc=261fe25c83a94fc3defe064baed3944cd3d16959;bpv=0;bpt=1;l=2203
- Folders named `go.mod` are ignored https://cs.opensource.google/go/go/+/master:src/cmd/go/internal/load/pkg.go;l=2158;drc=261fe25c83a94fc3defe064baed3944cd3d16959;bpv=0;bpt=1

I don't have the time available to implement this behavior correctly.

fmeum commented:
Thanks for collecting the pointers, that's already very helpful.

Given that you uncovered a number of places in which we (unknowingly) deviate from the native toolchain, let's just keep the scope of this PR limited to `all:` support. I certainly won't ask you to fix all of that.

How do you feel about doing the following as part of this PR?

1. Add excludes for ".bzr", ".hg", ".git", and ".svn", so that users of `all:` don't accidentally pick up their version control metadata when they wouldn't with the native Go toolchain.
2. Add the list in your last comment as a TODO to the code.

Let me know if you think that you won't have the time for it and we can consider merging as is.

linzhp commented:
> Do you have any backwards compatibility concerns?
No. I don't see a backward-compatibility issue here.

malt3 commented:
I implemented excludes for ".bzr", ".hg", ".git", and ".svn" and added TODOs.
I also noticed that the current implementation is already correct when it comes to the inclusion of hidden leaf nodes that I thought was not correct at first:

> Hidden files can be conditionally in-/excluded based on if the folder name is given with a glob or not
given the folder a and the file a/.hidden, the directive //go:embed a would exclude the file, while //go:embed a/* would include the file.

This works in the same way that it does for the native toolchain.

fmeum commented on go/tools/builders/embedcfg.go:
```diff
@@ -382,3 +414,23 @@ func fsValidPath(name string) bool {
+// can't or won't be included in modules and therefore shouldn't be treated
+// as existing for embedding.
+//
+// TODO: This should use the equivalent of golang.org/x/mod/module.CheckFilePath instead of fsValidPath.
```
```suggestion
// TODO: This should use the equivalent of golang.org/x/mod/module.CheckFilePath instead of fsValidPath.
// https://cs.opensource.google/go/go/+/master:src/cmd/go/internal/load/pkg.go;l=2200;drc=261fe25c83a94fc3defe064baed3944cd3d16959
```

fmeum commented on go/tools/builders/embedcfg.go:
```diff
@@ -314,12 +340,18 @@ func resolveEmbed(embed fileEmbed, root *embedNode) (matchedPaths, matchedFiles
+		// unless "all:" prefix was set.
 		// See golang/go#42328.
 		matchTreeErr := matchNode.walk(func(childRel string, childNode *embedNode) error {
+			// TODO: Should check that directories along path do not begin a new module
```
```suggestion
			// TODO: Should check that directories along path do not begin a new module
			// https://cs.opensource.google/go/go/+/master:src/cmd/go/internal/load/pkg.go;l=2158;drc=261fe25c83a94fc3defe064baed3944cd3d16959
```

fmeum commented on go/tools/builders/embedcfg.go:
```diff
@@ -295,17 +295,43 @@ func resolveEmbed(embed fileEmbed, root *embedNode) (matchedPaths, matchedFiles
 		}
+
+		// TODO: Should check that directories along path do not begin a new module
+		// (do not contain a go.mod).
```
```suggestion
		// (do not contain a go.mod).
		// https://cs.opensource.google/go/go/+/master:src/cmd/go/internal/load/pkg.go;l=2158;drc=261fe25c83a94fc3defe064baed3944cd3d16959
```

fmeum commented:
Thanks! I added the links to the comments. This is good to go now.
